libs/Database.js

```diff
@@ -1026,47 +1021,51 @@ class Database {
-   * by witnesses <= @lastVerifiedBlockNumber
-   * @param lastVerifiedBlockNumber
+   * Used by light nodes to cleanup (unneeded) blocks / transactions already verified
+   * by witnesses <= lastVerifiedBlockNumber - blocksToKeep
    * @returns {Promise<void>}
    */
-  async cleanupBlocks(lastVerifiedBlockNumber) {
+  async cleanupLightNode() {
     if (!this.lightNode) {
       return;
     }
-    // await this.chain.deleteMany({ _id: { $lte: lastVerifiedBlockNumber } }, { session: this.session });
+    const params = await this.findOne({ contract: 'witnesses', table: 'params', query: {} });
+    if (params && params.lastVerifiedBlockNumber) {
+      console.log(`cleaning up light node blocks and transactions`);
+      const cleanupUntilBlock = params.lastVerifiedBlockNumber - 1 - this.blocksToKeep;
+      await this.cleanupBlocks(cleanupUntilBlock);
+      await this.cleanupTransactions(cleanupUntilBlock);
+    }
+  }
 
-    await this.chain.deleteMany({ $and: [{ _id: { $gt: 0 } }, { _id: { $lte: lastVerifiedBlockNumber } }] }, { session: this.session });
+  /**
+   * Used by light nodes to cleanup (unneeded) blocks already verified
+   * by witnesses <= lastVerifiedBlockNumber - blocksToKeep
+   * @param cleanupUntilBlock cleanup blocks with a smaller blockNumber
+   * @returns {Promise<void>}
+   */
+  async cleanupBlocks(cleanupUntilBlock) {
+    await this.chain.deleteMany({ $and: [{ _id: { $gt: 0 } }, { _id: { $lte: cleanupUntilBlock } }] }, { session: this.session });
   }
 
   /**
    * Used by light nodes to cleanup (unneeded) transactions
+   * @param cleanupUntilBlock cleanup transactions with a smaller blockNumber
    * @returns {Promise<void>}
    */
-  async cleanupTransactions() {
-    if (!this.lightNode) {
-      return;
-    }
-    const transactionsTable = this.database.collection('transactions');
-    await transactionsTable.drop({ session: this.session });
-    await this.database.createCollection('transactions', { session: this.session });
+  async cleanupTransactions(cleanupUntilBlock) {
+    await this.database.collection('transactions').deleteMany({ blockNumber: { $lte: cleanupUntilBlock } }, { session: this.session });
   }
 
   /**
    * Checks if a node was a light node previously and returns true in case it was. Light nodes
-   * don't store transaction data, so the check simply checks if there are any transactions in
+   * drop transaction data after a certain time, so the check tries to get the first transaction. form
    * the database.
    * @returns {Promise<boolean>}
    */
   async wasLightNodeBefore() {
-    const transactionsTable = this.database.collection('transactions');
-    const someTransaction = await transactionsTable.findOne({}, { session: this.session });
-    if (someTransaction) {
-      // transaction table should be empty for a light node
-      return false;
-    }
-    return true;
+    const transaction = await this.getTransactionInfo(0);
```

Comment:
Not sure if this will work as intended.  getTransactionInfo expects a string tx ID, and just looking at the first few entries in the transactions table I see stuff like this:

![image](https://user-images.githubusercontent.com/8011332/172279793-efea5ce8-8baa-4673-aa24-21ba3cfdecd4.png)

I think it makes more sense to use `this.getBlockInfo(0)` for this check.

Reply:
Good point. But `this.getBlockInfo(0)` does not work either, as light nodes also store the genesis block (see comment above). If I update the code as suggested above, this would work - otherwise I would have to think of another solution.

Reply:
What about simply using `this.getBlockInfo(1)` ?